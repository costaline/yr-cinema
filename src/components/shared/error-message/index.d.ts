import React from 'react';

declare const ErrorMessage: React.FC;

export { ErrorMessage };
