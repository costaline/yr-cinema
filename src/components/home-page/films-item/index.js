export { default } from './films-item';
