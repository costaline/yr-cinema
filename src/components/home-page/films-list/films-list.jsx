import React from 'react';
import T from 'prop-types';
import styled from 'styled-components';

import FilmsItem from '~components/home-page/films-item';

export const FilmsList = ({ films }) => {
  return (
    <StyledSection>
      {films.map((film) => {
        const { name, posterURL } = film;

        const filmsItemProps = {
          name,
          posterURL
        };

        return <FilmsItem key={film.filmId} {...filmsItemProps} />;
      })}
    </StyledSection>
  );
};

FilmsList.propTypes = {
  films: T.arrayOf(
    T.shape({
      name: T.string,
      posterURL: T.string
    })
  )
};

const StyledSection = styled.section`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
`;
