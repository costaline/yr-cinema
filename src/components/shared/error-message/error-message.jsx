import React from 'react';

export const ErrorMessage = () => (
  <>
    <h1>Something went wrong...</h1>
  </>
);
