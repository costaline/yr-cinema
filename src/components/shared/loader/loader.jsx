import React from 'react';

export const Loader = () => {
  return <h2>Loading...</h2>;
};
