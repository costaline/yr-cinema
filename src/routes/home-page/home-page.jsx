import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import T from 'prop-types';
import queryString from 'query-string';
import styled from 'styled-components';

import { FilmsList } from '~components/home-page/films-list';
import { withErrorBoundary } from '~hocs/with-error-boundary';
import { fetchFilms } from '~store/films/actions';
import {
  getFilms,
  getLoading,
  getError,
  getTotal
} from '~store/films/selectors';
import { Loader } from '~components/shared/loader';
import { ErrorMessage } from '~components/shared/error-message';
import Pagination from '~components/shared/pagination/pagination';

const HomePage = ({ fetchFilms, films = [], total, loading, error }) => {
  // TODO: вынести определение текущей страницы в hoc
  const parsedQuery = queryString.parse(location.search);

  const currentPage = Number(parsedQuery.page) || 1;

  useEffect(() => {
    fetchFilms(currentPage);
  }, [currentPage]);

  return (
    <>
      <StyledContent>
        {loading && <Loader />}
        {error && <ErrorMessage />}
        {!loading && !error && <FilmsList films={films} />}
      </StyledContent>

      {total && (
        <Pagination total={total} limit={5} currentPage={currentPage} />
      )}
    </>
  );
};

HomePage.propTypes = {
  fetchFilms: T.func.isRequired,
  films: T.array,
  total: T.number,
  loading: T.bool.isRequired,
  error: T.shape({
    message: T.string
  })
};

const mapStateToProps = (state) => {
  return {
    films: getFilms(state),
    total: getTotal(state),
    loading: getLoading(state),
    error: getError(state)
  };
};

export default compose(
  connect(mapStateToProps, { fetchFilms }),
  withErrorBoundary
)(HomePage);

const StyledContent = styled.div`
  min-height: 475px;
  display: flex;
  justify-content: center;
  align-items: center;
`;
