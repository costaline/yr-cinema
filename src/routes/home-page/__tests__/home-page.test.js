import React from 'react';
import { shallow } from 'enzyme';

import { HomePage } from '../home-page';
import { Loader } from '~components/shared/loader';
import { ErrorMessage } from '~components/shared/error-message';
import { FilmsList } from '~components/home-page/films-list';
import Pagination from '~components/shared/pagination/pagination';

describe('Home Page with...', () => {
  const props = {
    fetchFilms: () => {},
    films: [],
    total: null,
    loading: false,
    error: null,
    location: {}
  };

  describe('Loader', () => {
    const nextProps = {
      ...props,
      loading: true
    };

    const wrapper = shallow(<HomePage {...nextProps} />);

    it('render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('contain', () => {
      expect(wrapper.contains(<Loader />)).toEqual(true);
    });
  });

  describe('Error message', () => {
    const nextProps = {
      ...props,
      error: 'Error message'
    };

    const wrapper = shallow(<HomePage {...nextProps} />);

    it('render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('contain', () => {
      expect(wrapper.contains(<ErrorMessage />)).toEqual(true);
    });
  });

  describe('Films list', () => {
    const nextProps = {
      ...props,
      films: [{ name: 'name1' }, { name: 'name1' }]
    };

    const wrapper = shallow(<HomePage {...nextProps} />);

    it('render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('contain', () => {
      expect(wrapper.find(FilmsList)).toHaveLength(1);
    });
  });

  describe('Pagination', () => {
    const nextProps = {
      ...props,
      total: 3
    };

    const wrapper = shallow(<HomePage {...nextProps} />);

    it('render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('contain', () => {
      expect(wrapper.find(Pagination)).toHaveLength(1);
    });
  });
});
