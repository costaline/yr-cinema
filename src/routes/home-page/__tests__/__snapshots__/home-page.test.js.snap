// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page with... Error message render correctly 1`] = `
<Fragment>
  <styled.div>
    <ErrorMessage />
  </styled.div>
</Fragment>
`;

exports[`Home Page with... Films list render correctly 1`] = `
<Fragment>
  <styled.div>
    <FilmsList
      films={
        Array [
          Object {
            "name": "name1",
          },
          Object {
            "name": "name1",
          },
        ]
      }
    />
  </styled.div>
</Fragment>
`;

exports[`Home Page with... Loader render correctly 1`] = `
<Fragment>
  <styled.div>
    <Loader />
  </styled.div>
</Fragment>
`;

exports[`Home Page with... Pagination render correctly 1`] = `
<Fragment>
  <styled.div>
    <FilmsList
      films={Array []}
    />
  </styled.div>
  <Memo(Pagination)
    currentPage={1}
    limit={5}
    sort={1}
    total={3}
  />
</Fragment>
`;
